File: curriculum/algos/quick_sort.py
def quick_sort(values):
	pending = [(0, len(values))]
	while pending:
		first, last = pending.pop()
		if last - first < 2:
			continue
		pivot = last - 1
		boundary = first
		for current in range(first, pivot):
			if values[current] < values[pivot]:
				values[current], values[boundary] = values[boundary], values[current]
				boundary += 1
		values[boundary], values[pivot] = values[pivot], values[boundary]
		pending.append((first, boundary))
		pending.append((boundary + 1, last))

def recursive_quick_sort(values, first=0, last=None):
	if last is None:
		last = len(values)
	if last - first < 2:
		return
	pivot = last - 1
	boundary = first
	for current in range(first, pivot):
		if values[current] < values[pivot]:
			values[current], values[boundary] = values[boundary], values[current]
			boundary += 1
	values[boundary], values[pivot] = values[pivot], values[boundary]
	recursive_quick_sort(values, first, boundary)
	recursive_quick_sort(values, boundary + 1, last)

File: curriculum/algos/quick_sort.rs
fn partition(values: &mut [i32], first: usize, last: usize) -> usize {
let pivot = last - 1;
let mut boundary = first;
for current in first..pivot {
if values[current] < values[pivot] {
values.swap(current, boundary);
boundary += 1;
}
}
values.swap(boundary, pivot);
boundary
}

pub fn quick_sort(values: &mut [i32]) {
let mut pending = vec![(0, values.len())];
while let Some((first, last)) = pending.pop() {
if last - first < 2 { continue; }
let boundary = partition(values, first, last);
pending.push((first, boundary));
pending.push((boundary + 1, last));
}
}

pub fn recursive_quick_sort(values: &mut [i32]) {
fn sort(values: &mut [i32], first: usize, last: usize) {
if last - first < 2 { return; }
let boundary = partition(values, first, last);
sort(values, first, boundary);
sort(values, boundary + 1, last);
}
sort(values, 0, values.len());
}
